Feature EPL-1501: Selectors (#205)
etendosoftware/docs: docs/developer-guide/etendo-classic/concepts/selectors.md

---
tags: 
  - selectors
  - etendo
  - hql
---

# Selectors

## Overview

In this section, the process to define selectors in the Etendo system is described.

Selectors combine a suggestion box with a popup window with filter and sorting capabilities. They have a number of important features:

>   * they can be completely defined without manual coding. 

Reviewer: ```suggestion
  * They can be completely defined without manual coding. 
```

>   * the selector definition (columns, search criteria, where clause) can be changed at runtime without re-compiling or re-starting the system. 

Reviewer: ```suggestion
  * The selector definition (columns, search criteria, where clause) can be changed at runtime without re-compiling or re-starting the system. 
```

>   * the selector makes it very easy to show linked information of a selected entity or search, sort and filter on linked information. 

Reviewer: ```suggestion
  * The selector makes it very easy to show linked information of a selected entity or search, sort and filter on linked information. 
```

>   * the selector suggestion box, as well as the popup grid, support paging with lookahead. This makes them also suitable for large datasets. 

Reviewer: ```suggestion
  * The selector suggestion box, as well as the popup grid, support paging with lookahead. This makes them also suitable for large datasets. 
```

## Selector Concept

A selector is defined within the Application Dictionary of Etendo. A selector is used for representing foreign key references in the Etendo user interface. A selector is defined as part of a Reference (a domain type) definition within Etendo.

A selector consists of three parts:

>   * the reference definition 

Reviewer: ```suggestion
  * The reference definition.
```

>   * the selector header defining which table to select from and the where clause to use 

Reviewer: ```suggestion
  * The selector header defining which table to select from and the where clause to use.
```

>   * the selector fields which define which columns are searched and what columns are shown in the popup window. 

Reviewer: ```suggestion
  * The selector fields which define which columns are searched and what columns are shown in the popup window.
```

## Defining a Selector

To define a new selector go to `Application Dictionary` > `Reference` window.
Defining the selector is a three-step procedure:

  1. create a new reference 
  2. create a selector header 
  3. create zero or more selector fields 

This section guides you through all three steps.

### Step 1: Define a Reference

The reference is used in the column definition. The definition is shown below in the screenshot.

![](../../../assets/developer-guide/etendo-classic/concepts/Selectors-2.png)

A description of the fields:

  * Name, Description and Help should be set to a specific value for the reference you are creating 
  * Base Reference **must** be set to false 

>   * Parent Reference **must** be set to OBUISEL_Selector_Reference 

Reviewer: ```suggestion
  * Parent Reference **must** be set to `OBUISEL_Selector_Reference` 
```

![](../../../assets/developer-guide/etendo-classic/concepts/Selectors-3.png)

A description of the fields:

  * **Name and Help/Comment**: set these to values describing the selector 
  * **Description**: this is used as the title of the popup window 
  * **Template**: there should normally be only one option: Selector Template. Multiple options are possible if additional templates have been installed for selectors. 
  * **Table**: normally a selector selects from a table, select the relevant table from the list 
  * **Column**: the column in the referenced table to which the foreign key column points. If nothing is set here (the default) then the primary key column is used. 
  * **Datasource**: can be set when no table is selected. A data source can be used to provide data which is not read from the database but is, for example, computed at runtime. For more information on Datasources, see the [Datasource](../concepts/datasources.md) Developers Manual.
  * **HQL Where Clause**: this where clause is used to filter the data read from the database. 

!!! Note
    Currently it is not possible to use session variables in the where clause.

!!! Note

>     When using properties of the entity in the whereclause always the prefix 'e.' has to be used, for example 'e.seqno'  

Reviewer: ```suggestion
    When using properties of the entity in the whereclause always the prefix `e.` has to be used, for example `e.seqno`  
```

  * when the data in a selector is not even read from the database but computed in memory by the system. 

A datasource is implemented by a developer and then defined in the Application Dictionary. After this, it can be used to define selectors.

!!!note
    For more information on how to implement a datasource, see its [Datasources](../concepts/datasources.md) page.

### Translated information

The Selector and Selector Fields also allow you to specify translated versions of names and titles shown in the user interface. To use translated names and titles, go to the **Defined Selector translation** and **Defined Selector Field Translation** tabs. These tabs are available at the top of the window when editing a Selector or Selector Field.

## Using a Selector

The next step is making use of the selector in the data model. The selector is the same as any other Reference in Etendo. The Reference is used in the definition of a Column. To use the defined selector and set it in a column, go to `Application Dictionary` > `Tables and Columns`. Then select the table and the column you want to set the reference for.

> Then, in the reference field, choose OBUISEL_Selector Reference. Then in the next combo box **Reference Search Key**, select the specific selector. In the screenshot below a Business Partner Selector is chosen.

Reviewer: ```suggestion
Then, in the reference field, choose `OBUISEL_Selector` Reference. Then in the next combo box **Reference Search Key**, select the specific selector. In the screenshot below a Business Partner Selector is chosen.
```

![](../../../assets/developer-guide/etendo-classic/concepts/Selectors-11.png)

The column can be used to define a field in the Etendo window in a standard way. So in the Etendo Window/Tab/Field definition there is no special setting or handling required for the selector.

After setting the reference in the column and using it in a specific window, you need to restart Etendo to see it in action.

## Changing a Selector at Runtime

The selector definition itself can be changed at runtime without re-compiling the system. There are no real limits to what part of the selector definition which can be changed at runtime, a non-excluding list:

>   * add/remove columns from the popup grid 

Reviewer: ```suggestion
  * Add/remove columns from the popup grid.
```

>   * change the where clause 

Reviewer: ```suggestion
  * Change the where clause.
```

>   * change the fields which are searched as part of the suggestion box display 

Reviewer: ```suggestion
  * Change the fields which are searched as part of the suggestion box display.
```

!!! note
    In order to change a selector, the module it belongs to needs to be in development.
    If not, then it is possible that the client-side javascript (representing the selector) is not refreshed automatically.

## Defining Out Fields

The usual return values from a Selector is the record **id** and **indentifier**. In some cases the developer wants to return more fields and not only these two. You can mark a Selector Field as **Out Field** and will be part of the JavaScript returned object.

![](../../../assets/developer-guide/etendo-classic/concepts/Selectors-13.png)

!!! note 
    By default the **id** and **identifier** fields are part of the returned object .

### Using Out Fields in Manual Code (Advanced)

> You can use the New Selectors in manual code (not generated by WAD). You must include all the necessary JS imports. There is an empty hook function **onValueChanged** that you can override in your manual window, to perform any custom action. When the user picks a row, this function gets executed. By default it does not do anything.

Reviewer: ```suggestion
You can use the New Selectors in manual code (not generated by WAD). You must include all the necessary JS imports. There is an empty hook function `onValueChanged` that you can override in your manual window, to perform any custom action. When the user picks a row, this function gets executed. By default it does not do anything.
```
